diff -Naur linux-3.9.4/drivers/staging/media/lirc/Kconfig linux-3.9.4.patch/drivers/staging/media/lirc/Kconfig
--- linux-3.9.4/drivers/staging/media/lirc/Kconfig	2013-05-24 20:45:59.000000000 +0200
+++ linux-3.9.4.patch/drivers/staging/media/lirc/Kconfig	2013-05-30 18:18:57.238957100 +0200
@@ -63,10 +63,17 @@
 	help
 	  Driver for the SIR IrDA port

+config LIRC_XBOX
+	tristate "XBOX USB IR Remote"
+	depends on LIRC && USB
+	help
+	  Driver for the Microsoft XBOX USB IR Remote
+
 config LIRC_ZILOG
 	tristate "Zilog/Hauppauge IR Transmitter"
 	depends on LIRC && I2C
 	help
 	  Driver for the Zilog/Hauppauge IR Transmitter, found on
 	  PVR-150/500, HVR-1200/1250/1700/1800, HD-PVR and other cards
+
 endif

