--- a/drivers/amlogic/video_dev/amlvideo.c	2017-02-01 10:10:34.000000000 +0100
+++ b/drivers/amlogic/video_dev/amlvideo.c	2017-02-08 11:04:38.334645058 +0100
@@ -476,8 +476,9 @@
 	strcpy(cap->card, "amlvideo");
 	strlcpy(cap->bus_info, dev->v4l2_dev.name, sizeof(cap->bus_info));
 	cap->version = AMLVIDEO_VERSION;
-	cap->capabilities = V4L2_CAP_VIDEO_CAPTURE | V4L2_CAP_STREAMING
+	cap->device_caps = V4L2_CAP_VIDEO_CAPTURE | V4L2_CAP_STREAMING
 				| V4L2_CAP_READWRITE;
+	cap->capabilities = cap->device_caps | V4L2_CAP_DEVICE_CAPS;
 	return 0;
 }
 
--- a/drivers/amlogic/video_dev/amlvideo2.c	2017-02-01 10:10:34.000000000 +0100
+++ b/drivers/amlogic/video_dev/amlvideo2.c	2017-02-08 11:07:02.446746257 +0100
@@ -3801,9 +3801,10 @@
 		node->vid_dev->v4l2_dev.name,
 		sizeof(cap->bus_info));
 	cap->version = AMLVIDEO2_VERSION;
-	cap->capabilities = V4L2_CAP_VIDEO_CAPTURE
+	cap->device_caps = V4L2_CAP_VIDEO_CAPTURE
 		|V4L2_CAP_STREAMING
 		| V4L2_CAP_READWRITE;
+	cap->capabilities = cap->device_caps | V4L2_CAP_DEVICE_CAPS;
 	return 0;
 }
 
