--- a/fs/compat_ioctl.c	2017-02-08 10:06:37.599814000 +0100
+++ b/fs/compat_ioctl.c	2017-02-08 10:26:12.739090000 +0100
@@ -945,7 +945,7 @@
  */
 #define XFORM(i) (((i) ^ ((i) << 27) ^ ((i) << 17)) & 0xffffffff)
 
-#define COMPATIBLE_IOCTL(cmd) XFORM(cmd),
+#define COMPATIBLE_IOCTL(cmd) XFORM(((long)cmd)),
 /* ioctl should not be warned about even if it's not implemented.
    Valid reasons to use this:
    - It is implemented with ->compat_ioctl on some device, but programs
