--- a/arch/arm64/boot/dts/amlogic/gxbb_p200.dts	2017-01-17 22:09:06.000000000 +0100
+++ b/arch/arm64/boot/dts/amlogic/gxbb_p200.dts	2017-02-02 10:32:07.559102393 +0100
@@ -52,12 +52,12 @@
 		};
 		pstore:aml_pstore {
 			compatible = "amlogic, pstore";
-			reg = <0x0 0x07300000 0x0 0x100000>;
+			reg = <0x0 0x20000000 0x0 0x100000>;
 			no-map;
 		};
 		fb_reserved:linux,meson-fb {
 			compatible = "amlogic, fb-memory";
-			size = <0x0 0x1900000>;
+			size = <0x0 0x2000000>;
 			no-map;
 		};
 
@@ -102,8 +102,7 @@
 		picdec_cma_reserved:linux,picdec {
 			compatible = "shared-dma-pool";
 			reusable;
-			size = <0x0 0x0>;
-			status = "disabled";
+			size = <0x0 0x3000000>;
 			alignment = <0x0 0x400000>;
 			linux,contiguous-region;
 		};
@@ -141,7 +140,7 @@
 		interrupts = <0 3 1
 			0 89 1>;
 		interrupt-names = "viu-vsync", "rdma";
-		mem_size = <0x01800000 0x00100000>; /* fb0/fb1 memory size */
+		mem_size = <0x01800000 0x00800000>; /* fb0/fb1 memory size */
 		display_mode_default = "1080p60hz";
 		scale_mode = <1>; /** 0:VPU free scale 1:OSD free scale 2:OSD super scale */
 		display_size_default = <1920 1080 1920 3240 32>; //1920*1080*4*3 = 0x17BB000
@@ -190,15 +189,15 @@
 			status = "okay";
 	};
 	ethmac: ethernet@0xc9410000{
-                compatible = "amlogic, gxbb-rgmii-dwmac";
+                compatible = "amlogic, gxbb-rmii-dwmac";
                 reg = <0x0 0xc9410000 0x0 0x10000
                 0x0 0xc8834540 0x0 0x8>;
                 interrupts = <0 8 1>;
-                phy-mode= "rgmii";
+                phy-mode= "rmii";
                 pinctrl-names = "eth_pins";
                 pinctrl-0 = <&eth_pins>;
                 rst_pin-gpios = <&gpio GPIOZ_14 0>;
-                mc_val = <0x1621>;
+                mc_val = <0x1800>;
                 resets = <&clock GCLK_IDX_ETHERNET>;
                 reset-names = "ethpower";
                 interrupt-names = "macirq";
@@ -206,23 +205,6 @@
                 clock-names = "ethclk81";
         };
 
-	pwm {
-		compatible = "amlogic, gx-pwm";
-		status = "okay";
-		pwm-outputs = <0>,<1>,<2>,<3>,<4>,<5>,<6>,<7>;
-		reg = <0x0 0xc1108550 0x0 0x30>,
-				<0x0 0xc8100550 0x0 0x10>;
-		clocks = <&clock CLK_XTAL>,
-				<&clock CLK_VID_PLL>,
-				<&clock CLK_FPLL_DIV4>,
-				<&clock CLK_FPLL_DIV3>;
-		clock-names = "xtal",
-				"vid_pll_clk",
-				"fclk_div4",
-				"fclk_div3";
-		clock-select = <0x0>,<0x0>,<0x0>,<0x0>,<0x3>,<0x0>,<0x0>,<0x0>;
-	};
-
 	mesonstream {
 		compatible = "amlogic, codec, streambuf";
 		dev_name = "mesonstream";
@@ -371,22 +353,13 @@
 			3: 250.0M    4: 333.3M    5: 400.0M
 			6: 500.0M    7: 666.7M */
 	};
-	bt-dev{
-		compatible = "amlogic, bt-dev";
-		dev_name = "bt-dev";
-		status = "ok";
-		gpio_reset = <&gpio       GPIOX_20       GPIO_ACTIVE_HIGH>;
-		//power_low_level;
-	};
-	bluesleep{
-		compatible = "amlogic, bluesleep";
-		dev_name = "bluesleep";
-		status = "ok";
-		interrupts = <  0 70 1>;
-		bt_port_id = < 1 >;
-		//bt_ext_wake = <&gpio       GPIOX_22       GPIO_ACTIVE_HIGH>;
-		bt_host_wake = <&gpio       GPIOX_21       GPIO_ACTIVE_HIGH>;
-	};
+//	bt-dev{
+//		compatible = "amlogic, bt-dev";
+//		dev_name = "bt-dev";
+//		status = "ok";
+//		gpio_reset = <&gpio       GPIOX_20       GPIO_ACTIVE_HIGH>;
+//		//power_low_level;
+//	};
 
 	rtc{
 		compatible = "amlogic, aml_vrtc";
@@ -397,14 +370,14 @@
 	};
 
 	wifi{
-		compatible = "amlogic, aml_wifi";
-		dev_name = "aml_wifi";
+		compatible = "amlogic, aml_broadcm_wifi";
+		dev_name = "aml_broadcm_wifi";
 		status = "okay";
 		interrupt_pin = <&gpio       GPIOX_7       GPIO_ACTIVE_HIGH>;
 		interrupts = <	0 68 4>;
-		irq_trigger_type = "GPIO_IRQ_LOW";
+		irq_trigger_type = "GPIO_IRQ_HIGH";
 		power_on_pin = <&gpio       GPIOX_6       GPIO_ACTIVE_HIGH>;
-		dhd_static_buf;
+		power_on_pin2 = <&gpio       GPIOX_20       GPIO_ACTIVE_HIGH>;
 		pinctrl-names = "wifi_32k_pins";
 		pinctrl-0 = <&wifi_32k_pins>;
 	};
@@ -480,7 +453,7 @@
 		dev_name = "aml_newsd.0";
 		status = "okay";
 		reg = <0x0 0xd0070000 0x0 0x2000>;
-		interrupts = <	0 216 1>;
+		interrupts = <	0 216 4>;
 		pinctrl-names =  "sdio_clk_cmd_pins", "sdio_all_pins";
 		pinctrl-0 = <&sdio_clk_cmd_pins>;
 		pinctrl-1 = <&sdio_all_pins>;
@@ -871,9 +844,7 @@
 		port-speed = <0>; /** 0: default, high, 1: full */
 		port-config = <0>; /** 0: default */
 		port-dma = <0>; /** 0: default, 1: single, 2: incr, 3: incr4, 4: incr8, 5: incr16, 6: disable*/
-		port-id-mode = <1>; /** 0: hardware, 1: sw_host, 2: sw_slave*/
-		gpio-vbus-power = "GPIODV_24";
-		gpios = <&gpio       GPIODV_24       GPIO_ACTIVE_HIGH>;
+		port-id-mode = <0>; /** 0: hardware, 1: sw_host, 2: sw_slave*/
 		gpio-work-mask	= <1>; /**0: work on pulldown,1:work on pullup*/
 		phy-reg = <0xc0000000>;
 		phy-reg-size = <0x20>;
@@ -1137,11 +1108,12 @@
 			};
 		};
 	};
-
-/*
 	dvb {
-		compatible = "amlogic, dvb";
+		compatible = "amlogic,dvb";
 		dev_name = "dvb";
+		status = "okay";
+		fec_reset_gpio-gpios = <&gpio GPIOY_13 GPIO_ACTIVE_HIGH>;
+		power_ctrl_gpio-gpios = <&gpio GPIOY_11 GPIO_ACTIVE_LOW>;
 		//"parallel","serial","disable"
 		ts0 = "parallel";
 		ts0_control = <0>;
@@ -1156,28 +1128,33 @@
 		reset-names = "demux", "asyncfifo", "ahbarb0", "uparsertop";
 	};
 	dvbfe {
-		compatible = "amlogic, dvbfe";
-		dev_name = "dvbfe";
-		dtv_demod0 = "Mxl241";
-		dtv_demod0_i2c_adap_id = <2>;
-		dtv_demod0_i2c_addr = <0x60>;
-		dtv_demod0_reset_value = <0>;
-		dtv_demod0_reset_gpio-gpios = <&gpio GPIOY_12 GPIO_ACTIVE_HIGH>;
+		compatible = "amlogic,dvbfe";
+		//dev_name = "avl6211";
+		status = "okay";
+		dtv_demod0 = "Avl6211";
+		dtv_demod0_i2c_adap_id = <1>;
+		dtv_demod0_i2c_addr = <0xC0>;
+ 		dtv_demod0_reset_value = <0>;
+		//dtv_demod0_reset_gpio = <&gpio GPIOY_13 GPIO_ACTIVE_HIGH>;
+		dtv_demod0_reset_gpio-gpios = <&gpio GPIOY_13 GPIO_ACTIVE_LOW>;
+		dtv_demod0_power_gpio-gpios = <&gpio GPIOY_11 GPIO_ACTIVE_LOW>;
 		fe0_dtv_demod = <0>;
 		fe0_ts = <0>;
 		fe0_dev = <0>;
 	};
-*/
 };
 &i2c_a {
-  status = "disabled";
-  /*p200: multiplex with usb PWR, disbaled*/
+  status = "okay";
 };
-/*
 &i2c_b {
   status = "okay";
 };
-*/
+&i2c_c {
+  status = "okay";
+};
+&i2c_d {
+  status = "okay";
+};
 
 &i2c_slave {
 	status = "disabled";
